### DISCUSSION TITLE
how to retrieve definition across modules / packages programatically using lsp client?

### DISCUSSION THREAD

**Discussion by code2graph** [python-lsp/python-lsp-server#422] [Q&A]
how to code up a lsp client to find references programatically?

I know as **python-lsp-server**  is a language server and theoreticallty I can connect to it and extract references from a function call is invoked.

For example for the following code:

```
import os
from getpass import getpass
from GPT4Readability.utils import *
import importlib.resources as pkg_resources  


def generate_readme(root_dir, output_name, model):
    """Generates a README.md file based on the python files in the provided directory

    Args:
        root_dir (str): The root directory of the python package to parse and generate a readme for
    """

    # prompt_folder_name = os.path.join(os.path.dirname(__file__), "prompts")
    # prompt_path = os.path.join(prompt_folder_name, "readme_prompt.txt")

    with pkg_resources.open_text('GPT4Readability.prompts','readme_prompt.txt') as f:         
        inb_msg = f.read()

    # with open(prompt_path) as f:
    #     lines = f.readlines()
    # inb_msg = "".join(lines)

    file_check_result = check_files_in_directory(root_dir)

```

Here I need to find out `check_files_in_directory` is defined in the `GPT4Readability.utils`.

I am struggling to find this informaiton and any help is welcome.

**Comment by smheidrich** [comment]
As you said yourself and as the name "Language _Server_ Protocol" implies, you'll need a client that connects to the language server in order to make use of it programmatically.

You might want to use a client library for this, e.g. for Python [pylspclient](https://github.com/yeger00/pylspclient) seems reasonable (but I haven't tried it & unfortunately the code seems the be the only documentation). That library in particular has a [`definition`](https://github.com/yeger00/pylspclient/blob/d662a0bfecb0416fbc0bff3ff946491a3de8ada4/pylspclient/lsp_client.py#L237-L243) method to make a "Goto Definition Request" to the language server, which would then return the data necessary to do so.
